client/asset/btc/spv.go

```diff
@@ -0,0 +1,1631 @@
+		}
+		return msgTx.TxOut[vout], confs, nil
+	}
+	if !errors.Is(err, WalletTransactionNotFound) {
+		return nil, 0, fmt.Errorf("walletTransaction error: %w", err)
+	}
+
+	var blockHash *chainhash.Hash
+	if txDetails != nil && txDetails.Block.Hash != (chainhash.Hash{}) {
+		blockHash = &txDetails.Block.Hash
+	}
+
+	// We don't really know if it's spent, so we'll need to scan.
+	utxo, err := w.scanFilters(txHash, vout, pkScript, startTime, blockHash)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if utxo == nil || utxo.spend != nil {
+		return nil, 0, nil
+	}
+
+	if utxo.blockHash == nil {
+		return nil, 0, fmt.Errorf("output %s:%v not found", txHash, vout)
+	}
+
+	tip, err := w.cl.BestBlock()
+	if err != nil {
+		return nil, 0, fmt.Errorf("BestBlock error: %v", err)
+	}
+
+	confs := uint32(tip.Height) - utxo.blockHeight + 1
+
+	return utxo.txOut, confs, nil
+}
+
+// filterScanFromHeight scans BIP158 filters beginning at the specified block
+// height until the tip, or until a spending transaction is found.
+func (w *spvWallet) filterScanFromHeight(txHash chainhash.Hash, vout uint32, pkScript []byte, startBlockHeight int32, checkPt *filterScanResult) (*filterScanResult, error) {
+	tip, err := w.getBestBlockHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	res := checkPt
+	if res == nil {
+		res = new(filterScanResult)
+	}
+
+search:
+	for height := startBlockHeight; height <= tip; height++ {
+		if res.spend != nil && res.blockHash == nil {
+			w.log.Debugf("A spending input was found during the scan but the output " +
+				"itself wasn't found. Was the startBlockHeight early enough?")
+			return res, nil
+		}
+		blockHash, err := w.getBlockHash(int64(height))
+		if err != nil {
+			return nil, fmt.Errorf("error getting block hash for height %d: %w", height, err)
+		}
+		res.checkpoint = *blockHash
+		matched, err := w.matchPkScript(blockHash, [][]byte{pkScript})
+		if err != nil {
+			return nil, fmt.Errorf("matchPkScript error: %w", err)
+		}
+		if !matched {
+			continue search
+		}
+		// Pull the block.
+		block, err := w.cl.GetBlock(*blockHash)
+		if err != nil {
+			return nil, fmt.Errorf("GetBlock error: %v", err)
+		}
+		msgBlock := block.MsgBlock()
+
+		// Scan every transaction.
+	nextTx:
+		for _, tx := range msgBlock.Transactions {
+			// Look for a spending input.
+			if res.spend == nil {
+				for vin, txIn := range tx.TxIn {
+					prevOut := &txIn.PreviousOutPoint
+					if prevOut.Hash == txHash && prevOut.Index == vout {
+						res.spend = &spendingInput{
+							txHash:      tx.TxHash(),
+							vin:         uint32(vin),
+							blockHash:   *block.Hash(),
+							blockHeight: uint32(height),
+						}
+						if res.blockHash != nil {
+							break search
+						}
+						// The output could still be in this block, just not
+						// in this transaction.
+						continue nextTx
+					}
+				}
+			}
+			// Only check for the output if this is the right transaction.
+			if tx.TxHash() != txHash {
```

Comment:
Not sure but maybe `if res.blockHash != nil ||  tx.TxHash() != txHash {` is faster because you can skip `tx.TxHash()` sometimes. Maybe doesn't matter.